src/server/scripts/EasternKingdoms/Gnomeregan/boss_mekgineer_thermaplugg.cpp

```diff
@@ -0,0 +1,464 @@
+
+            _events.Update(diff);
+
+            while (uint32 eventId = _events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                case EVENT_POUND:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 5.0f))
+                        DoCast(target, SPELL_POUND);
+                    else
+                        DoCastVictim(SPELL_POUND);
+
+                    _events.RescheduleEvent(EVENT_POUND, 10 * IN_MILLISECONDS);
+                    _events.RescheduleEvent(EVENT_STEAM_BLAST, 4 * IN_MILLISECONDS);
+                    break;
+                case EVENT_STEAM_BLAST:
+                {
+                    uint8 rand = urand(1, 4);
+                    if (rand == 1)
+                        Talk(SAY_MACHINES);
+                    DoCastVictim(SPELL_STEAM_BLAST);
+                    if (GetThreat(me->GetVictim()))
+                        ModifyThreatByPercent(me->GetVictim(), -50);
+                }
+                break;
+                case EVENT_ACTIVATE_BOMBS:
+                {
+                    uint32 faceEntry = Trinity::Containers::SelectRandomContainerElement(_availableFacesList);
+                    if (GameObject* face = me->FindNearestGameObject(faceEntry, 125.0f))
+                    {
+                        Talk(SAY_EXPLOSIONS);
+                        face->AI()->DoAction(ACTION_ACTIVATE);
+                        face->SetGoState(GO_STATE_ACTIVE);
+                        DoCast(SPELL_ACTIVATE_BOMB_VISUAL);
+                    }
+
+                    _events.RescheduleEvent(EVENT_ACTIVATE_BOMBS, urand(10 * IN_MILLISECONDS, 18 * IN_MILLISECONDS));
+                }
+                break;
+                case EVENT_WELDING_BEAM:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 30, true))
+                        DoCast(target, SPELL_WELDING_BEAM, true);
+
+                    _events.RescheduleEvent(EVENT_WELDING_BEAM, urand(10 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
+                    break;
+                default:
+                    break;
+                }
+            }
+
+            DoMeleeAttackIfReady();
+        }
+
+        EventMap _events;
+        std::set<uint32> _availableFacesList;
+    };
+
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return GetGnomereganAI<boss_mekgineer_thermapluggAI>(creature);
+    }
+
+};
+
+class npc_walking_bomb : public CreatureScript
+{
+public:
+    npc_walking_bomb() : CreatureScript("npc_walking_bomb") { }
+
+    struct npc_walking_bombAI : public ScriptedAI
+    {
+        npc_walking_bombAI(Creature* creature) : ScriptedAI(creature)
+        {
+            _instance = me->GetInstanceScript();
+            me->ApplySpellImmune(0, IMMUNITY_STATE, SPELL_AURA_MOD_TAUNT, true);
+            me->ApplySpellImmune(0, IMMUNITY_EFFECT, SPELL_EFFECT_ATTACK_ME, true);
+            me->SetReactState(REACT_PASSIVE);
```

Comment:
I don't like this in the ctor, put it in `InitializeAI` just to avoid it being changed afterwards